txpipe/source_selector.py

```diff
@@ -0,0 +1,431 @@
+        group.create_dataset('R_total', (nbin_source,2,2), dtype='f')
+
+        return outfile
+
+    def write_tomography(self, outfile, start, end, source_bin, R):
+        """
+        Write out a chunk of tomography and response.
+
+        Parameters
+        ----------
+
+
+        outfile: h5py.File
+
+        start: int
+            The index into the output this chunk starts at
+
+        end: int
+            The index into the output this chunk ends at
+
+        tomo_bin: array of shape (nrow,)
+            The bin index for each output object
+
+        R: array of shape (nrow,2,2)
+            Multiplicative bias calibration factor for each object
+
+
+        """
+        group = outfile['tomography']
+        group['source_bin'][start:end] = source_bin
+        group = outfile['metacal_response']
+        group['R_gamma'][start:end,:,:] = R
+
+    def write_global_values(self, outfile, calibrators, number_density_stats):
+        """
+        Write out overall selection biases
+
+        Parameters
+        ----------
+
+        outfile: h5py.File
+
+        S: array of shape (nbin,2,2)
+            Selection bias matrices
+        """
+        nbin_source = len(calibrators)
+
+        R = np.zeros((nbin_source, 2, 2))
+        S = np.zeros((nbin_source, 2, 2))
+        N = np.zeros(nbin_source)
+        R_scalar = np.zeros(nbin_source)
+
+        # this needs fixing
+        sigma_e = number_density_stats.collect()
+
+        for i, cal in enumerate(calibrators):
+            R[i], S[i], N[i] = cal.collect(self.comm)
+            sigma_e[i] /= 0.5*(R[i,0,0] + R[i,1,1])
+        
+
+        if self.rank==0:
+            group = outfile['metacal_response']
+            group['R_S'][:,:,:] = S
+            group['R_gamma_mean'][:,:,:] = R
+            group['R_total'][:,:,:] = R + S
+            group = outfile['tomography']
+            group['sigma_e'][:] = sigma_e
+            # These are the same in metacal
+            group['source_counts'][:] = N
+            group['N_eff'][:] = N
+
+    
+    def select(self, data, bin_index):
+        s2n_cut = self.config['s2n_cut']
+        T_cut = self.config['T_cut']
+        verbose = self.config['verbose']
+
+        s2n = data['mcal_s2n']
+        T = data['mcal_T']
+        zbin = data['zbin']
+
+        Tpsf = data['mcal_psf_T_mean']
+        flag = data['mcal_flags']
+
+        n0 = len(flag)
+        sel  = flag==0
+        f1 = sel.sum() / n0
+        sel &= (T/Tpsf)>T_cut
+        f2 = sel.sum() / n0
+        sel &= s2n>s2n_cut
+        f3 = sel.sum() / n0
+        sel &= zbin==bin_index
+        f4 = sel.sum() / n0
+        
+        variant = data.suffix
+        if verbose:
+            print(f"Bin {bin_index} ({variant}) {f1:.2%} flag, {f2:.2%} size, {f3:.2%} SNR, {f4:.2%} z")
+
+        return sel
+
+
+def flatten_list(lst):
+    return [item for sublist in lst for item in sublist]
+
+
+if __name__ == '__main__':
+    PipelineStage.main()
+
```

Comment:
Can you let me know on the PR page if you made any changes to this, other than splitting it off into its own file?

Reply:
don't think so, yeah I'll add to the PR page